### PR TITLE
Add clang-format support

### PR DESCRIPTION
**Pull request by MathiasMagnus**
This work if the continuation of #889.

## Notes for the reviewers

Some comments from the original work addressed here, most notably:

- Limit column size of all formatted sources to [100](https://github.com/KhronosGroup/KTX-Software/pull/889#discussion_r1573654200).

Aside from explicit comments:

- A git-blame ignore file has been committed to the repo which instructs GitHub's blame functionality on the web interface to keep attribution of lines affected by formatting. Local clones may be [configured told similarly](https://www.stefanjudis.com/today-i-learned/how-to-exclude-commits-from-git-blame/).
- The functional changes are all outside the singular "Apply initial formatting" commit.

Should other formatting changes need refining, discussions may continue here.

### CODE REVIEW

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -25,7 +25,11 @@ on:
   workflow_dispatch:
 
 jobs:
+  formatting:
+    uses: ./.github/workflows/formatting.yml
   windows:
+    # Shortcircuit and don't burn CI time when formatting will reject
+    needs: formatting
```
Comment (MarkCallow) [inline .github/workflows/windows.yml:32]: Is this forcing CI to always run formatting before a build? If so, how? The workflow name is "Formatting".

File: .git-blame-ignore-revs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Copyright 2024 The Khronos Group Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+# Apply initial formatting
+4ceab17e94e73c918130a49ad1b21d5e4c14e7ec
```
Comment (MarkCallow) [inline .git-blame-ignore-revs:5]: Isn't this ignoring just this single commit? What about future reformatting that takes place when commit(s) are pushed?

File: interface/basisu_c_binding/inc/basisu_c_binding.h [hunk 3]
```diff
@@ -4,36 +4,32 @@
 #pragma once
 
 #if defined(KHRONOS_STATIC)
-  #define KTX_BASISU_API
+#define KTX_BASISU_API
```
Comment (MarkCallow) [inline interface/basisu_c_binding/inc/basisu_c_binding.h:7]: I prefer to keep the PP directives indented. I see from fmt's .clang_format that it is possible:

`IndentPPDirectives: BeforeHash`

will do it.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (MathiasMagnus) [reply .github/workflows/windows.yml:32]: The workflow name is "Formatting" indeed, but that is only the display name of the workflow if it were to run by itself, which it doesn't, because the only `on` flavor is [`on.workflow_call`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_call), which means only other workflows will call it (aka. [reusable worfklow](https://docs.github.com/en/actions/using-workflows/reusing-workflows)). The calling workflow will give it a new name (if it doesn't, it'll be an ugly yml filename). The `needs` part has to refer to to the [`jobs.<job_id>`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_id) of the workflow being called in the _caller's_ yaml file.

For an example of how such a workflow executes, check [this](https://github.com/MathiasMagnus/Test-Packaging/actions/runs/9414364871) run from a toy repo of mine. Hope this clears it up.

File: .git-blame-ignore-revs (hunk 2 above)
Comment (MathiasMagnus) [reply .git-blame-ignore-revs:5]: Thanks for reminding me, this commit hash needs update after rebases I done.

Ideally no further commits need to be added here, because now the entire repo is formatted and all formatting violations will trip CI, hence no pure formatting-related commits need to be made. Contributors are highly encouraged to configure their development environment to format on save. (eg. VS Code has [`editor.formatOnSave`](https://code.visualstudio.com/docs/editor/codebasics#_formatting) option and multiple C/C++ formatting extensions available, most notably [ms-vscode.cpptools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) or [llvm-vs-code-extensions.vscode-clangd](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd))

File: interface/basisu_c_binding/inc/basisu_c_binding.h (hunk 3 above)
Comment (MathiasMagnus) [reply interface/basisu_c_binding/inc/basisu_c_binding.h]: Added.

**Comment by MathiasMagnus** [comment]
I think this work is complete. The current CI failure doesn't seem to be related to the changes made here.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (MarkCallow) [reply .github/workflows/windows.yml:32]: I think I see. `needs:` refers to the job name `formatting` within the windows.yml file. Correct? The workflow name given in formatting.yml is unused here. Correct?

Comment (MathiasMagnus) [reply .github/workflows/windows.yml:32]: Yes.

File: .git-blame-ignore-revs (hunk 2 above)
Comment (MarkCallow) [reply .git-blame-ignore-revs]: Please resolve this when you've updated the commit hash.

Comment (MathiasMagnus) [reply .git-blame-ignore-revs]: It's been updated with the latest rebase.

File: external/.clang-format [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Copyright 2024 The Khronos Group Inc.
+# SPDX-License-Identifier: Apache-2.0
+---
+# Disable clang-format in this directory
+DisableFormat: true
+SortIncludes: false
```
Comment (MarkCallow) [inline external/.clang-format:6]: Why is this needed when formatting is disabled?

File: interface/basisu_c_binding/src/basisu_c_binding.cpp [hunk 5]
```diff
@@ -150,56 +137,53 @@ uint32_t basis_file::getImageTranscodedSizeInBytes(uint32_t image_index, uint32_
...
 
-        if (transcoder_format == transcoder_texture_format::cTFPVRTC1_4_RGB || transcoder_format == transcoder_texture_format::cTFPVRTC1_4_RGBA)
-        {
-            // For PVRTC1, Basis only writes (or requires) total_blocks * bytes_per_block. But GL requires extra padding for very small textures: 
+        if (transcoder_format == transcoder_texture_format::cTFPVRTC1_4_RGB ||
+            transcoder_format == transcoder_texture_format::cTFPVRTC1_4_RGBA) {
+            // For PVRTC1, Basis only writes (or requires) total_blocks * bytes_per_block. But GL
+            // requires extra padding for very small textures:
             // https://www.khronos.org/registry/OpenGL/extensions/IMG/IMG_texture_compression_pvrtc.txt
             // The transcoder will clear the extra bytes followed the used blocks to 0.
```
Comment (MarkCallow) [inline interface/basisu_c_binding/src/basisu_c_binding.cpp:188]: Unrelated to this PR but given the overhead of running a PR please fix the typo. s/followed/following/.

File: lib/astc_encode.cpp [hunk 6]
```diff
@@ -35,19 +35,19 @@
...
-#include <pthread.h>
+    #include <pthread.h>
 #else
 // Provide pthreads support on windows
-#define WIN32_LEAN_AND_MEAN
-#include <Windows.h>
+    #define WIN32_LEAN_AND_MEAN
+    #include <Windows.h>
 
 typedef HANDLE pthread_t;
```
Comment (MarkCallow) [inline lib/astc_encode.cpp:44]: Can you get it to indent this code up to the #endif?

File: tests/loadtests/geom/cube.h [hunk 7]
```diff
@@ -3,70 +3,47 @@
  * SPDX-License-Identifier: Apache-2.0
```
Comment (MarkCallow) [inline tests/loadtests/geom/cube.h:3]: Either don't format this file or fix the formatting for it so the vertices remain separated in the cube face - the formatting has made the comments nonsensical plus it is much harder to see the vertices when they are all run together.

File: tests/loadtests/geom/cube_data.h [hunk 8]
```diff
@@ -10,146 +10,92 @@
 // Mesh and VertexFormat Data
```
Comment (MarkCallow) [inline tests/loadtests/geom/cube_data.h:10]: Also don't format this.

File: tests/loadtests/geom/frame.h [hunk 9]
```diff
@@ -6,18 +6,6 @@
 /*
```
Comment (MarkCallow) [inline tests/loadtests/geom/frame.h:6]: Or this.

File: tests/loadtests/geom/frame.h [hunk 10]
```diff
@@ -6,18 +6,6 @@
...
  * Frame definition
```
Comment (MarkCallow) [inline tests/loadtests/geom/frame.h:7]: Or this.

File: tests/loadtests/geom/quad.h [hunk 11]
```diff
@@ -9,26 +9,10 @@
 /*
```
Comment (MarkCallow) [inline tests/loadtests/geom/quad.h:9]: Or this.

File: tests/unittests/unittests.cc [hunk 12]
```diff
@@ -85,31 +84,30 @@ class CheckHeader1Test : public ::testing::Test {
...
     KTX_header testHeader;
+
   public:
     static ktx_uint8_t ktxId[12];
 };
 
-ktx_uint8_t CheckHeader1Test::ktxId[12] = {
-    0xAB, 0x4B, 0x54, 0x58, 0x20, 0x31, 0x31, 0xBB, 0x0D, 0x0A, 0x1A, 0x0A
-};
+ktx_uint8_t CheckHeader1Test::ktxId[12] = {0xAB, 0x4B, 0x54, 0x58, 0x20, 0x31,
```
Comment (MarkCallow) [inline tests/unittests/unittests.cc:92]: Is there a way to fix the format of a few lines in a file? This represents the signature of a KTX v1 file so it much easier for humans to follow as originally written.

File: .git-blame-ignore-revs (hunk 2 above)
Comment (MarkCallow) [reply .git-blame-ignore-revs]: Sorry. I edited my original comment that said "please resolve" to add the request about documentation which you probably missed. Probably `BUILDING.md` is the best place for the documentation.

**Comment by MarkCallow** [comment]
You need to rebase or merge `main` again to pick up the workaround for the build issue on GitHub Actions. The latest runner image is only partially fixed.

**Comment by MarkCallow** [comment]
@MathiasMagnus please address the outstanding issues and complete this work.

File: .git-blame-ignore-revs (hunk 2 above)
Comment (MathiasMagnus) [reply .git-blame-ignore-revs]: Documented VS Code. I looked into (Neo)vim, but the phase space is too large.

File: external/.clang-format (hunk 4 above)
Comment (MathiasMagnus) [reply external/.clang-format:6]: It's probably not strictly necessary.

**Comment by MathiasMagnus** [comment]
@MarkCallow We created this PR out of courtesy, because this task did not fit in the original project timeline, so we are unable to address all the specific personal styling and tooling requests. I added documentation on the minimal requirement to reduce CI friction and noted how to configure VS Code to not trip CI. Vim is too deep a rabbit a hole and I don't have the time to give anything I document in XCode a test drive.

As for the rest of the outstanding formatting related inquiries: instead of whack-a-mole-ing the formatting issues one-by-one and aligning them to the desired format, I reverted the "Initial formatting" commit and just commit the machinery enabling the enforced formatting.

- For now, CI job definition is there, but its invocation is commented out.
- The GitHub-friendly git blame ignore file is there with a note but not hash.

Because clang-format can't run on an entire source tree, all the tools drive it file-by-file. Should you want to tweak the `.clang-format` files to perfection, I used the following one-liner on Windows:

```pwsh
# list folders excluding build and .venv | recurse into all | filter files matching regex | for each of the matches { run clang-format inplace using style files and full path to file }
gci -exc build,.venv | gci -rec | where -prop Name -match '^.*\.((((c|C)(c|pp|xx|\+\+)?$)|((h|H)h?(pp|xx|\+\+)?$))|(ino|pde|proto|cu))$' | % { clang-format.exe -i --style=file $_.FullName }
```

I leave implementing similar logic in the shell of their choice to the end-user. (The regex I borrowed from the [script run by the CI action](https://github.com/jidicula/clang-format-action/blob/4dacb6087962f49cedccf872a80c610db1e0e02a/check.sh#L77).)

In summary, all the infrastructure for clang-format is there, please tweak it for your needs as you see fit.

**Comment by MarkCallow** [comment]
I am not happy ...

> We created this PR out of courtesy, because this task did not fit in the original project timeline, so we are unable to address all the specific personal styling and tooling requests.

I thought providing a PR with initial styling was part of the contract. How else would you deliver the project work, if not by a PR?

> Instead of whack-a-mole-ing the formatting issues one-by-one and aligning them to the desired format, I reverted the "Initial formatting" commit and just commit the machinery enabling the enforced formatting.

There is no whack-a-mole on formatting issues here. The related open conversations are everything. I count 2 format change requests and requests to exclude 4 files from formatting. That is not a huge amount of work.

Please restore the formatting making the small number of requested changes.

I understand that documenting this for vim and xcode is potentially time-consuming, if you don't already have the knowledge. Omitting that part  of my request is okay.

**Comment by aqnuep** [comment]
> I thought providing a PR with initial styling was part of the contract. How else would you deliver the project work, if not by a PR?

Yes, but the contract ended at the end of March with the understanding that due to the additional requests we ran out of time and budget to do the clang-format change. We simply offered to put together a PR anyway afterwards, out of courtesy, as we had free time for it, but we cannot go back-and-forth for months to make the styling fit anybody's personal preferences.

All the infrastructure work is there. Actual styling is subjective and a usual rabbit-hole. You should easily tweak the configurations for your needs, and continuing this back-and-forth is not something that we can continue doing forever.

File: lib/astc_encode.cpp (hunk 6 above)
Comment (aqnuep) [reply lib/astc_encode.cpp]: You can exclude lines of code from being formatted by surrounding them with:

```cpp
// clang-format off
...
// clang-format on
```

There are examples of this in the generated files.

File: external/.clang-format (hunk 4 above)
Comment (aqnuep) [reply external/.clang-format:6]: This is the convention used in other Khronos repositories, but it should have no effects (at least on clang-format versions I've used).

File: tests/loadtests/geom/cube.h (hunk 7 above)
Comment (aqnuep) [reply tests/loadtests/geom/cube.h]: There is no way to exclude individual files from formatting. That's why we had to move everything external that is not expected to be formatted by us into a separate subdirectory where we disabled formatting.

You can exclude lines of code from being formatted (or entire files) by surrounding them with:

```cpp
// clang-format off
...
// clang-format on
```

There are examples of this in the generated files.

File: tests/loadtests/geom/cube_data.h (hunk 8 above)
Comment (aqnuep) [reply tests/loadtests/geom/cube_data.h]: See above.

File: tests/loadtests/geom/frame.h (hunk 9 above)
Comment (aqnuep) [reply tests/loadtests/geom/frame.h]: See above.

File: tests/loadtests/geom/quad.h (hunk 11 above)
Comment (aqnuep) [reply tests/loadtests/geom/quad.h]: See above.

File: tests/unittests/unittests.cc (hunk 12 above)
Comment (aqnuep) [reply tests/unittests/unittests.cc]: You can exclude lines of code from being formatted by surrounding them with:

```cpp
// clang-format off
...
// clang-format on
```

There are examples of this in the generated files.

**Comment by aqnuep** [comment]
> There is no whack-a-mole on formatting issues here. The related open conversations are everything. I count 2 format change requests and requests to exclude 4 files from formatting. That is not a huge amount of work.

Please also consider that rebasing a formatting change is a nightmare. We've undone the one-shot formatting, you can adjust the styling as needed, and re-run clang-format on all C++ files then merge it once you're happy with the output. It should be trivial from this point.

**Comment by MarkCallow** [comment]
> the understanding that due to the additional requests we ran out of time and budget to do the clang-format change

Were these "additional requests" related or unrelated to clang-format? If the former what where they?

> There is no way to exclude individual files from formatting. 

> Because clang-format can't run on an entire source tree, all the tools drive it file-by-file.

These statements conflict. What am I missing?

**Comment by MarkCallow** [comment]
> There is no way to exclude individual files from formatting.

What about .clang-format-ignore which I've just discovered at https://clang.llvm.org/docs/ClangFormat.html.

**Comment by aqnuep** [comment]
> > the understanding that due to the additional requests we ran out of time and budget to do the clang-format change
> 
> Were these "additional requests" related or unrelated to clang-format? If the former what where they?
> > There is no way to exclude individual files from formatting.
> 
> > Because clang-format can't run on an entire source tree, all the tools drive it file-by-file.
> 
> These statements conflict. What am I missing?
